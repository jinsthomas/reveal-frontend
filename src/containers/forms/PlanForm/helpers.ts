import { ErrorMessage, Field, FieldArray, Form, Formik, FormikErrors } from 'formik';
import { omit, pick } from 'lodash';
import moment from 'moment';
import React, { FormEvent } from 'react';
import { Button, FormGroup, Label } from 'reactstrap';
import * as Yup from 'yup';
import DatePickerWrapper from '../../../components/DatePickerWrapper';
import {
  DATE_FORMAT,
  DEFAULT_ACTIVITY_DURATION_DAYS,
  DEFAULT_PLAN_DURATION_DAYS,
} from '../../../configs/env';
import {
  actionReasons,
  FIClassifications,
  FIReasons,
  FIStatuses,
  goalPriorities,
  planActivities,
  PlanActivity,
} from '../../../configs/settings';
import { DATE, IRS_TITLE, IS, NAME, REQUIRED, SAVING } from '../../../constants';
import { InterventionType, PlanStatus } from '../../../store/ducks/plans';

/** separate FI and IRS activities */
export const FIActivities = omit(planActivities, ['IRS']);
export const IRSActivities = pick(planActivities, ['IRS']);

/** Allowed FI Status values */
export type FIStatusType = typeof FIStatuses[number];

/** Allowed FI Status values */
export type FIReasonType = typeof FIReasons[number];

/** Array of FI Statuses */
export const fiStatusCodes = Object.values(FIClassifications).map(e => e.code as FIStatusType);

/** Yup validation schema for PlanForm */
export const PlanSchema = Yup.object().shape({
  activities: Yup.array().of(
    Yup.object().shape({
      actionDescription: Yup.string().required(REQUIRED),
      actionReason: Yup.string()
        .oneOf(Object.values(actionReasons))
        .required(REQUIRED),
      actionTitle: Yup.string().required(REQUIRED),
      goalDescription: Yup.string().required(REQUIRED),
      goalDue: Yup.date().required(REQUIRED),
      goalPriority: Yup.string()
        .oneOf(Object.values(goalPriorities))
        .required(REQUIRED),
      goalValue: Yup.number()
        .min(1)
        .required(REQUIRED),
      timingPeriodEnd: Yup.date().required(REQUIRED),
      timingPeriodStart: Yup.date().required(REQUIRED),
    })
  ),
  caseNum: Yup.string(),
  date: Yup.string().required(`${DATE} ${IS} ${REQUIRED}`),
  end: Yup.date().required(REQUIRED),
  fiReason: Yup.string().oneOf(FIReasons.map(e => e)),
  fiStatus: Yup.string().oneOf(fiStatusCodes),
  interventionType: Yup.string()
    .oneOf(Object.values(InterventionType))
    .required(REQUIRED),
  name: Yup.string().required(`${NAME} ${IS} ${REQUIRED}`),
  opensrpEventId: Yup.string(),
  start: Yup.date().required(REQUIRED),
  status: Yup.string()
    .oneOf(Object.values(PlanStatus))
    .required(REQUIRED),
  title: Yup.string().required(REQUIRED),
});

/** Plan activity form fields interface */
export interface PlanActivityFormFields {
  actionDescription: string;
  actionReason: string;
  actionTitle: string;
  goalDescription: string;
  goalDue: Date;
  goalPriority: string;
  goalValue: number;
  timingPeriodEnd: Date;
  timingPeriodStart: Date;
}

/** Plan form fields interface */
export interface PlanFormFields {
  activities: PlanActivityFormFields[];
  caseNum?: string;
  date: Date;
  end: Date;
  fiReason?: FIReasonType;
  fiStatus?: FIStatusType;
  interventionType: InterventionType;
  name: string;
  opensrpEventId?: string;
  start: Date;
  status: PlanStatus;
  title: string;
}

/**
 * Convert a plan activity object to an object that can be used in the PlanForm
 * activities section
 * @param activityObj - the plan activity object
 */
export function extractActivityForForm(activityObj: PlanActivity): PlanActivityFormFields {
  const initialGoalDue = activityObj.goal.target[0].due;
  return {
    actionDescription: activityObj.action.description || '',
    actionReason: activityObj.action.reason || '',
    actionTitle: activityObj.action.title || '',
    goalDescription: activityObj.goal.description || '',
    goalDue:
      initialGoalDue && initialGoalDue !== ''
        ? moment(initialGoalDue).toDate()
        : moment()
            .add(DEFAULT_ACTIVITY_DURATION_DAYS, 'days')
            .toDate(),
    goalPriority: activityObj.goal.priority || goalPriorities[1],
    goalValue: activityObj.goal.target[0].detail.detailQuantity.value || 0,
    timingPeriodEnd:
      activityObj.action.timingPeriod.end && activityObj.action.timingPeriod.end !== ''
        ? moment(activityObj.action.timingPeriod.end).toDate()
        : moment()
            .add(DEFAULT_ACTIVITY_DURATION_DAYS, 'days')
            .toDate(),
    timingPeriodStart:
      activityObj.action.timingPeriod.start && activityObj.action.timingPeriod.start !== ''
        ? moment(activityObj.action.timingPeriod.start).toDate()
        : moment().toDate(),
  };
}

/**
 * Converts a plan activities objects to a list of activities for use on PlanForm
 * @param items - plan activities
 */
export function getFormActivities(items: typeof FIActivities | typeof IRSActivities) {
  return Object.values(items)
    .sort((a, b) => a.action.prefix - b.action.prefix)
    .map(e => extractActivityForForm(e));
}

/**
 * Get the plan name and title
 * @param {any} event - the event object
 * @param {PlanFormFields} formValues - the form values
 * @returns {[string, string]} - the plan name and title
 */
export const getNameTitle = (event: FormEvent, formValues: PlanFormFields): [string, string] => {
  const target = event.target as HTMLInputElement;
  let name = IRS_TITLE;
  let title = IRS_TITLE;
  const currentInterventionType =
    target.name === 'interventionType' ? target.value : formValues.interventionType;
  const currentFiStatus = target.name === 'fiStatus' ? target.value : formValues.fiStatus;
  const currentJurisdiction = 'Some Jurisdiction';
  const currentDate = target.name === 'date' ? target.value : formValues.date;
  if (currentInterventionType === InterventionType.FI) {
    const result = [
      currentFiStatus,
      currentJurisdiction,
      moment(currentDate).format(DATE_FORMAT.toUpperCase()),
    ].map(e => {
      if (e) {
        return e;
      }
    });
    name = result.join('-');
    title = result.join(' ');
  }
  return [name, title];
};

/**
 * Check if field in field array has an error
 * @param field - the form field
 * @param index - the index in the field array
 * @param errors - the list of error objects
 */
export function doesFieldHaveErrors(
  field: string,
  index: number,
  errors: Array<FormikErrors<PlanActivityFormFields> | undefined>
) {
  return (
    errors && Object.entries(errors[index] || {}).filter(([key, val]) => key === field).length > 0
  );
}
