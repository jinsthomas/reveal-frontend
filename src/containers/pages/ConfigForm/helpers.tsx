import React from 'react';
import Loading from '../../../components/page/Loading';
import { OPENSRP_API_BASE_URL } from '../../../configs/env';
import { growl } from '../../../helpers/utils';
import { getPayloadOptions } from '../../../services/opensrp';

export const defaultConfigProps = {
  LoadingComponent: <Loading />,
  baseURL: OPENSRP_API_BASE_URL,
  getPayload: getPayloadOptions,
  growl,
};
