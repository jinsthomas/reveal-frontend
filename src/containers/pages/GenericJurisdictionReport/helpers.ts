import { Cell, Column } from 'react-table';
<<<<<<< HEAD:src/containers/pages/GenericJurisdictionReport/helpers.ts
import { getIRSThresholdAdherenceIndicator } from '../../../helpers/indicators';
=======
import { getIRSThresholdAdherenceIndicator } from '../../../../helpers/indicators';
>>>>>>> Replacing typings and props dependent on react-table-v6:src/containers/pages/IRS/JurisdictionsReport/helpers.ts

/** columns for Namibia IRS jurisdictions */
export const NamibiaColumns = [
  {
    Header: 'Name',
    columns: [
      {
        accessor: 'jurisdiction_name',
        minWidth: 180,
      },
    ],
  },
  {
    Header: 'Structures Targeted',
    columns: [
      {
        accessor: 'target_2019',
      },
    ],
  },
  {
    Header: 'Structures Found',
    columns: [
      {
        accessor: 'structuresfound',
      },
    ],
  },
  {
    Header: 'Structures Sprayed',
    columns: [
      {
        accessor: 'structuressprayed',
      },
    ],
  },
  {
    Header: 'Target Coverage',
    columns: [
      {
        Cell: (cell: Cell) => getIRSThresholdAdherenceIndicator(cell),

        accessor: 'targetcoverage',
        className: 'indicator centered',
      },
    ],
  },
  {
    Header: 'Found Coverage',
    columns: [
      {
        Cell: (cell: Cell) => getIRSThresholdAdherenceIndicator(cell),

        accessor: 'foundcoverage',
        className: 'indicator centered',
      },
    ],
  },
  {
    Header: 'Refusals',
    columns: [
      {
        Cell: (cell: Cell) => getIRSThresholdAdherenceIndicator(cell, null),
        Header: 'Following first visit',
        accessor: 'refusalsfirst',
        className: 'indicator centered',
        id: 'refusalsfirst',
      },
      {
        Cell: (cell: Cell) => getIRSThresholdAdherenceIndicator(cell, null),
        Header: 'Following mop-up',
        accessor: 'refusalsmopup',
        className: 'indicator centered',
        id: 'refusalsmopup',
      },
    ],
    headerClassName: 'centered',
  },
  {
    Header: 'Locked',
    columns: [
      {
        Cell: (cell: Cell) => getIRSThresholdAdherenceIndicator(cell, null),
        Header: 'Following first visit',
        accessor: 'lockedfirst',
        className: 'indicator centered',
        id: 'lockedfirst',
      },
      {
        Cell: (cell: Cell) => getIRSThresholdAdherenceIndicator(cell, null),
        Header: 'Following mop-up',
        accessor: 'lockedmopup',
        className: 'indicator centered',
        id: 'lockedmopup',
      },
    ],
    headerClassName: 'centered',
  },
];

/** columns for Zambia IRS jurisdictions */
export const ZambiaJurisdictionsColumns = [
  {
    Header: 'Name',
    columns: [
      {
        Header: '',
        accessor: 'jurisdiction_name',
        minWidth: 180,
      },
    ],
  },
  {
    Header: 'Total Spray Areas',
    columns: [
      {
        Header: '',
        accessor: 'totareas',
      },
    ],
  },
  {
    Header: 'Targeted Spray Areas',
    columns: [
      {
        Header: '',
        accessor: 'targareas',
      },
    ],
  },
  {
    Header: '% visited Spray Areas Effectively sprayed',
    columns: [
      {
        Cell: (cell: Cell) => getIRSThresholdAdherenceIndicator(cell),
        Header: '',
        accessor: 'perctvisareaseffect',
        className: 'indicator centered',
      },
    ],
  },
  {
    Header: 'Total Structures',
    columns: [
      {
        Header: '',
        accessor: 'totstruct',
      },
    ],
  },
  {
    Header: 'Targeted Structures',
    columns: [
      {
        Header: '',
        accessor: 'targstruct',
      },
    ],
  },
  {
    Header: 'Sprayed Structures',
    columns: [
      {
        Header: '',
        accessor: 'sprayedstruct',
      },
    ],
  },
  {
    Header: 'Spray coverage of targeted (Progress)',
    columns: [
      {
        Cell: (cell: Cell) => getIRSThresholdAdherenceIndicator(cell),
        Header: '',
        accessor: 'spraycovtarg',
        className: 'indicator centered',
      },
    ],
  },
  {
    Header: 'Room coverage of structures sprayed',
    columns: [
      {
        Cell: (cell: Cell) => getIRSThresholdAdherenceIndicator(cell),
        Header: '',
        accessor: 'roomcov',
        className: 'indicator centered',
      },
    ],
  },
];

/** columns for  mda point jurisdictions */
export const mdaJurisdictionsColumns = [
  {
    Header: 'Name',
    columns: [
      {
        Header: '',
        accessor: 'jurisdiction_name',
        minWidth: 180,
      },
    ],
  },
  {
    Header: 'Total SACs Registered',
    columns: [
      {
        Header: '',
        accessor: 'sacregistered',
      },
    ],
  },
  {
    Header: 'MMA Coverage',
    columns: [
      {
        Header: '',
        accessor: 'mmacov',
      },
    ],
  },
  {
    Header: 'MMA Coverage (%)',
    columns: [
      {
        Cell: (cell: CellInfo) => getIRSThresholdAdherenceIndicator(cell),
        Header: '',
        accessor: 'mmacovper',
      },
    ],
  },
  {
    Header: 'SACs Refused',
    columns: [
      {
        Header: '',
        accessor: 'sacrefused',
      },
    ],
  },
  {
    Header: 'SACs Sick/Pregnant/Contraindicated',
    columns: [
      {
        Header: '',
        accessor: 'sacrefmedreason',
      },
    ],
  },
  {
    Header: 'ADR Reported (%)',
    columns: [
      {
        Cell: (cell: CellInfo) => getIRSThresholdAdherenceIndicator(cell),
        Header: '',
        accessor: 'mmaadr',
      },
    ],
  },
  {
    Header: 'ADR Severe (%)',
    columns: [
      {
        Cell: (cell: CellInfo) => getIRSThresholdAdherenceIndicator(cell),
        Header: '',
        accessor: 'mmaadrsev',
      },
    ],
  },
  {
    Header: 'Alb Tablets Distributed',
    columns: [
      {
        Header: '',
        accessor: 'albdist',
      },
    ],
  },
];

/** columns for Namibia IRS focus (spray) areas */
export const ZambiaFocusAreasColumns = [
  {
    Header: 'Name',
    columns: [
      {
        Header: '',
        accessor: 'jurisdiction_name',
        minWidth: 180,
      },
    ],
  },
  {
    Header: 'Structures on the ground',
    columns: [
      {
        Header: '',
        accessor: 'totstruct',
      },
    ],
  },
  {
    Header: 'Found',
    columns: [
      {
        Header: '',
        accessor: 'foundstruct',
      },
    ],
  },
  {
    Header: 'Sprayed',
    columns: [
      {
        Header: '',
        accessor: 'sprayedstruct',
      },
    ],
  },
  {
    Header: 'Spray Coverage (Effectiveness)',
    columns: [
      {
        Cell: (cell: Cell) => getIRSThresholdAdherenceIndicator(cell),
        Header: '',
        accessor: 'spraycov',
        className: 'indicator centered',
      },
    ],
  },
  {
    Header: 'Found Coverage',
    columns: [
      {
        Cell: (cell: Cell) => getIRSThresholdAdherenceIndicator(cell),
        Header: '',
        accessor: 'spraytarg',
        className: 'indicator centered',
      },
    ],
  },
  {
    Header: 'Spray Success Rate (PMI SC)',
    columns: [
      {
        Cell: (cell: Cell) => getIRSThresholdAdherenceIndicator(cell),
        Header: '',
        accessor: 'spraysuccess',
        className: 'indicator centered',
      },
    ],
  },
  {
    Header: 'Room coverage of structures sprayed',
    columns: [
      {
        Cell: (cell: Cell) => getIRSThresholdAdherenceIndicator(cell),
        Header: '',
        accessor: 'roomcov',
        className: 'indicator centered',
      },
    ],
  },
];

/** IRS Table Columns
 * These are all the table columns for IRS that we know about.
 */
export const plansTableColumns: { [key: string]: Column[] } = {
  mdaJurisdictionsColumns,
  namibia2019: NamibiaColumns,
  zambiaFocusArea2019: ZambiaFocusAreasColumns,
  zambiaJurisdictions2019: ZambiaJurisdictionsColumns,
};
