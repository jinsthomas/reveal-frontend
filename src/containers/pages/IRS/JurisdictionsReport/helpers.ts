import { CellInfo } from 'react-table';
import { getIRSThresholdAdherenceIndicator } from '../../../../helpers/indicators';
import { FlexObject } from '../../../../helpers/utils';

/** columns for Namibia IRS jurisdictions */
export const NamibiaColumns = [
  {
    Header: 'Name',
    columns: [
      {
        Header: '',
        accessor: 'jurisdiction_name',
        minWidth: 180,
      },
    ],
  },
  {
    Header: 'Structures Targeted',
    columns: [
      {
        Header: '',
        accessor: 'target_2019',
      },
    ],
  },
  {
    Header: 'Structures Found',
    columns: [
      {
        Header: '',
        accessor: 'structuresfound',
      },
    ],
  },
  {
    Header: 'Structures Sprayed',
    columns: [
      {
        Header: '',
        accessor: 'structuressprayed',
      },
    ],
  },
  {
    Header: 'Target Coverage',
    columns: [
      {
        Cell: (cell: CellInfo) => getIRSThresholdAdherenceIndicator(cell),
        Header: '',
        accessor: 'targetcoverage',
        className: 'indicator centered',
      },
    ],
  },
  {
    Header: 'Found Coverage',
    columns: [
      {
        Cell: (cell: CellInfo) => getIRSThresholdAdherenceIndicator(cell),
        Header: '',
        accessor: 'foundcoverage',
        className: 'indicator centered',
      },
    ],
  },
  {
    Header: 'Refusals',
    columns: [
      {
        Header: 'Following first visit',
        accessor: 'refusalsfirst',
        className: 'centered',
        id: 'refusalsfirst',
      },
      {
        Header: 'Following mop-up',
        accessor: 'refusalsmopup',
        className: 'centered',
        id: 'refusalsmopup',
      },
    ],
  },
  {
    Header: 'Locked',
    columns: [
      {
        Header: 'Following first visit',
        accessor: 'lockedfirst',
        className: 'centered',
        id: 'lockedfirst',
      },
      {
        Header: 'Following mop-up',
        accessor: 'lockedmopup',
        className: 'centered',
        id: 'lockedmopup',
      },
    ],
  },
];

/** columns for Zambia IRS jurisdictions */
export const ZambiaJurisdictionsColumns = [
  {
    Header: 'Name',
    columns: [
      {
        Header: '',
        accessor: 'jurisdiction_name',
        minWidth: 180,
      },
    ],
  },
  {
    Header: 'Total Spray Areas',
    columns: [
      {
        Header: '',
        accessor: 'totareas',
      },
    ],
  },
  {
    Header: 'Targeted Spray Areas',
    columns: [
      {
        Header: '',
        accessor: 'targareas',
      },
    ],
  },
  {
    Header: '% visited Spray Areas Effectively sprayed',
    columns: [
      {
        Cell: (cell: CellInfo) => getIRSThresholdAdherenceIndicator(cell),
        Header: '',
        accessor: 'perctvisareaseffect',
        className: 'indicator centered',
      },
    ],
  },
  {
    Header: 'Total Structures',
    columns: [
      {
        Header: '',
        accessor: 'totstruct',
      },
    ],
  },
  {
    Header: 'Targeted Structures',
    columns: [
      {
        Header: '',
        accessor: 'targstruct',
      },
    ],
  },
  {
    Header: 'Sprayed Structures',
    columns: [
      {
        Header: '',
        accessor: 'sprayedstruct',
      },
    ],
  },
  {
    Header: 'Spray coverage of targeted (Progress)',
    columns: [
      {
        Cell: (cell: CellInfo) => getIRSThresholdAdherenceIndicator(cell),
        Header: '',
        accessor: 'spraycovtarg',
        className: 'indicator centered',
      },
    ],
  },
  {
    Header: 'Room coverage of structures sprayed',
    columns: [
      {
        Cell: (cell: CellInfo) => getIRSThresholdAdherenceIndicator(cell),
        Header: '',
        accessor: 'roomcov',
        className: 'indicator centered',
      },
    ],
  },
];

/** columns for Namibia IRS focus (spray) areas */
export const ZambiaFocusAreasColumns = [
  {
    Header: 'Name',
    columns: [
      {
        Header: '',
        accessor: 'jurisdiction_name',
        minWidth: 180,
      },
    ],
  },
  {
    Header: 'Structures on the ground',
    columns: [
      {
        Header: '',
        accessor: 'totstruct',
      },
    ],
  },
  {
    Header: 'Found',
    columns: [
      {
        Header: '',
        accessor: 'foundstruct',
      },
    ],
  },
  {
    Header: 'Sprayed',
    columns: [
      {
        Header: '',
        accessor: 'sprayedstruct',
      },
    ],
  },
  {
    Header: 'Spray Coverage (Effectiveness)',
    columns: [
      {
        Cell: (cell: CellInfo) => getIRSThresholdAdherenceIndicator(cell),
        Header: '',
        accessor: 'spraycov',
        className: 'indicator centered',
      },
    ],
  },
  {
    Header: 'Found Coverage',
    columns: [
      {
        Cell: (cell: CellInfo) => getIRSThresholdAdherenceIndicator(cell),
        Header: '',
        accessor: 'spraytarg',
        className: 'indicator centered',
      },
    ],
  },
  {
    Header: 'Spray Success Rate (PMI SC)',
    columns: [
      {
        Cell: (cell: CellInfo) => getIRSThresholdAdherenceIndicator(cell),
        Header: '',
        accessor: 'spraysuccess',
        className: 'indicator centered',
      },
    ],
  },
  {
    Header: 'Room coverage of structures sprayed',
    columns: [
      {
        Cell: (cell: CellInfo) => getIRSThresholdAdherenceIndicator(cell),
        Header: '',
        accessor: 'roomcov',
        className: 'indicator centered',
      },
    ],
  },
];

/** extract hierarchical tree of jurisdictions as a list, with the highest
 * level item in the hierarchy as the 0th element
 * @param {FlexObject[]} tree - the data set
 * @param {FlexObject} item - the item in question
 * @returns {FlexObject[]} - the hierarchy for the item in question
 */
export function getTree(
  tree: FlexObject[],
  item: FlexObject,
  result: FlexObject[] = []
): FlexObject[] {
  if (item.jurisdiction_parent_id === null && !result.includes(item)) {
    result.push(item);
  } else {
    const parentItems = tree.filter(el => el.jurisdiction_id === item.jurisdiction_parent_id);
    if (parentItems.length > 0) {
      if (!result.includes(parentItems[0])) {
        result.push(parentItems[0]);
      }
      return getTree(tree, parentItems[0], result);
    }
  }
  return result;
}
