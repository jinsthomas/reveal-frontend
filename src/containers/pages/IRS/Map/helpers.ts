import GeojsonExtent from '@mapbox/geojson-extent';
import { BLACK, GREY, TASK_GREEN, TASK_ORANGE, TASK_RED, TASK_YELLOW } from '../../../../colors';
import { GisidaProps } from '../../../../components/GisidaWrapper';
import { circleLayerConfig, fillLayerConfig, lineLayerConfig } from '../../../../configs/settings';
import { MAIN_PLAN, STRUCTURE_LAYER } from '../../../../constants';
import { FlexObject } from '../../../../helpers/utils';
import { IRSJurisdiction } from '../../../../store/ducks/IRS/jurisdictions';
import { StructureFeatureCollection } from '../../../../store/ducks/IRS/structures';
import { Jurisdiction } from '../../../../store/ducks/jurisdictions';

/** Default indicator stops */
const defaultIndicatorStops = [
  ['Complete', TASK_GREEN],
  ['Not Sprayed', TASK_RED],
  ['Partially Sprayed', TASK_ORANGE],
  ['Not Visited', TASK_YELLOW],
  ['Not Eligible', BLACK],
];

/** Utility function to build structure layer
 * @param {StructureFeatureCollection} structures - Feature Collection of structures
 * @param {string[][]} indicatorStops - the indicator stops
 */
export const structuresLayerBuilder = (
  structures: StructureFeatureCollection,
  indicatorStops: string[][] = defaultIndicatorStops
) => {
  const structuresLayers: FlexObject[] = [];
  const layerType = structures.features[0].geometry && structures.features[0].geometry.type;
  const structuresPopup: FlexObject = {
    body: `<div>
          <p class="heading">{{structure_type}}</p>
          <p>Status: {{business_status}}</p>
        </div>`,
    join: ['structure_jurisdiction_id', 'structure_jurisdiction_id'],
  };

  const structureStatusColors = {
    default: GREY,
    property: 'business_status',
    stops: indicatorStops,
    type: 'categorical',
  };

  if (layerType === 'Point') {
    // build circle layers if structures are points
    const structureCircleLayer = {
      ...circleLayerConfig,
      id: `${STRUCTURE_LAYER}-circle`,
      paint: {
        ...circleLayerConfig.paint,
        'circle-color': structureStatusColors,
        'circle-radius': ['interpolate', ['linear'], ['zoom'], 13, 2, 14, 4, 16, 8, 20, 12],
        'circle-stroke-color': structureStatusColors,
        'circle-stroke-opacity': 1,
      },
      popup: structuresPopup,
      source: {
        ...circleLayerConfig.source,
        data: {
          data: JSON.stringify(structures),
          type: 'stringified-geojson',
        },
        type: 'geojson',
      },
      visible: true,
    };
    structuresLayers.push(structureCircleLayer);
  } else {
    // build fill / line layers if structures are polygons
    const structuresFillLayer = {
      ...fillLayerConfig,
      id: `${STRUCTURE_LAYER}-fill`,
      paint: {
        ...fillLayerConfig.paint,
        'fill-color': structureStatusColors,
        'fill-outline-color': structureStatusColors,
      },
      popup: structuresPopup,
      source: {
        ...fillLayerConfig.source,
        data: {
          ...fillLayerConfig.source.data,
          data: JSON.stringify(structures),
        },
      },
      visible: true,
    };
    structuresLayers.push(structuresFillLayer);

    const structuresLineLayer = {
      ...lineLayerConfig,
      id: `${STRUCTURE_LAYER}-line`,
      paint: {
        'line-color': structureStatusColors,
        'line-opacity': 1,
        'line-width': 2,
      },
      source: {
        ...lineLayerConfig.source,
        data: {
          ...lineLayerConfig.source.data,
          data: JSON.stringify(structures),
        },
      },
    };
    structuresLayers.push(structuresLineLayer);
  }
  return structuresLayers;
};

/** Get Gisida Wrapper Props
 * @param {Jurisdiction} jurisdiction - the jurisdiction (with geojson field)
 * @param {StructureFeatureCollection} structures - Feature Collection of structures
 * @param {string[][]} indicatorStops - the indicator stops
 */
export const getGisidaWrapperProps = (
  jurisdiction: Jurisdiction,
  structures: StructureFeatureCollection | null,
  indicatorStops: string[][] = defaultIndicatorStops
): GisidaProps | null => {
  if (!jurisdiction.geojson) {
    return null;
  }
  const layers: FlexObject[] = [];

  // define line layer for Jurisdiction outline
  const jurisdictionLineLayer = {
    ...lineLayerConfig,
    id: `${MAIN_PLAN}-${jurisdiction.jurisdiction_id}`,
    source: {
      ...lineLayerConfig.source,
      data: {
        ...lineLayerConfig.source.data,
        data: JSON.stringify(jurisdiction.geojson),
      },
    },
    visible: true,
  };
  layers.push(jurisdictionLineLayer);

  // Define structures layers
  let structuresLayers: FlexObject[] = [];
  if (structures) {
    structuresLayers = structuresLayerBuilder(structures, indicatorStops);
  }

  // define feature collection of all geoms being rendered
  const featureCollection =
    !structures || !structures.features || !structures.features.length
      ? { features: [jurisdiction.geojson], type: 'FeatureCollection' }
      : {
          ...structures,
          features: [...structures.features, jurisdiction.geojson],
        };
  // define bounds for gisida map position
  const bounds = GeojsonExtent(featureCollection);

  for (const structureLayer of structuresLayers) {
    layers.push(structureLayer);
  }

  const gisidaWrapperProps: GisidaProps = {
    bounds,
    geoData: null,
    handlers: [],
    layers,
    pointFeatureCollection: null,
    polygonFeatureCollection: null,
    structures: null,
  };
  return gisidaWrapperProps;
};

/** Get breadcrumbs for a jurisdiction object
 * This uses the jurisdiction_name_path and jurisdiction_path fields to get
 * the breadcrumbs of the parents of a given jurisdiction object
 * @param {IRSJurisdiction} jurisdiction - the jurisdiction in question
 * @param {string} urlPath - the url path that we append the jurisdiction id to
 */
export const getJurisdictionBreadcrumbs = (
  jurisdiction: IRSJurisdiction,
  urlPath: string = '/'
) => {
  const result = [];

  for (let i = 0; i < jurisdiction.jurisdiction_name_path.length; i++) {
    let url = urlPath;
    if (jurisdiction.jurisdiction_path[i]) {
      url = `${urlPath}/${jurisdiction.jurisdiction_path[i]}`;
    }
    result.push({
      label: jurisdiction.jurisdiction_name_path[i],
      url,
    });
  }

  return result;
};
