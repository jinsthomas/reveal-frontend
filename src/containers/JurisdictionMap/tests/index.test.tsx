import reducerRegistry from '@onaio/redux-reducer-registry';
import { mount, shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import flushPromises from 'flush-promises';
import React from 'react';
import { Provider } from 'react-redux';
import { MemoryRouter } from 'react-router';
import * as errorHelpers from '../../../helpers/errors';
import store from '../../../store';
import jurisdictionReducer, {
  fetchJurisdictions,
  reducerName as jurisdictionReducerName,
} from '../../../store/ducks/jurisdictions';
import { jurisdiction1 } from '../../../store/ducks/tests/fixtures';
import { AN_ERROR_OCCURRED } from '../.././../configs/lang';
import ConnectedJurisdictionMap, {
  defaultProps,
  JurisdictionMap,
  JurisdictionMapProps,
} from '../index';

/** register the jurisdictions reducer */
reducerRegistry.register(jurisdictionReducerName, jurisdictionReducer);

jest.mock('../../../configs/env');

describe('containers/JurisdictionMap', () => {
  beforeEach(() => {
    jest.resetAllMocks();
  });

  it('renders without crashing', () => {
    const props: JurisdictionMapProps = {
      ...defaultProps,
      jurisdiction: jurisdiction1,
      jurisdictionId: jurisdiction1.jurisdiction_id,
    };
    shallow(
      <MemoryRouter>
        <JurisdictionMap {...props} />
      </MemoryRouter>
    );
  });

  it('renders correctly', async () => {
    const callbackMock: any = jest.fn();
    const supersetServiceMock: any = jest.fn();
    supersetServiceMock.mockImplementation(async () => []);

    const props: JurisdictionMapProps = {
      ...defaultProps,
      callback: callbackMock,
      cssClass: 'super-custom',
      jurisdiction: jurisdiction1,
      jurisdictionId: jurisdiction1.jurisdiction_id,
      supersetService: supersetServiceMock,
    };
    const wrapper = mount(
      <MemoryRouter>
        <JurisdictionMap {...props} />
      </MemoryRouter>
    );

    await flushPromises();
    wrapper.update();

    const wrapperDiv = wrapper.find('div.super-custom');
    expect(wrapperDiv.length).toEqual(1);

    expect(supersetServiceMock.mock.calls).toEqual([
      [
        1 /** this comes from the envs mock */,
        {
          adhoc_filters: [
            {
              clause: 'WHERE',
              comparator: '450fc15b-5bd2-468a-927a-49cb10d3bcac',
              expressionType: 'SIMPLE',
              operator: '==',
              subject: 'jurisdiction_id',
            },
          ],
          row_limit: 1,
        },
      ],
    ]);
    expect(supersetServiceMock).toHaveBeenCalledTimes(1);

    expect(callbackMock.mock.calls).toEqual([[jurisdiction1]]);
    expect(callbackMock).toHaveBeenCalledTimes(1);

    expect(wrapper.find('GisidaWrapper').props()).toEqual({
      currentGoal: null,
      geoData: jurisdiction1,
      goal: null,
      handlers: [],
      minHeight: '200px',
      pointFeatureCollection: null,
      polygonFeatureCollection: null,
      structures: null,
    });
  });

  it('renders fetch error correctly', async () => {
    const mockDisplayError: any = jest.fn();
    (errorHelpers as any).displayError = mockDisplayError;

    const supersetServiceMock: any = jest.fn();
    supersetServiceMock.mockImplementation(async () => undefined);

    const props: JurisdictionMapProps = {
      ...defaultProps,
      jurisdictionId: jurisdiction1.jurisdiction_id,
      supersetService: supersetServiceMock,
    };
    const wrapper = mount(
      <MemoryRouter>
        <JurisdictionMap {...props} />
      </MemoryRouter>
    );

    await flushPromises();
    wrapper.update();

    expect(mockDisplayError.mock.calls).toEqual([[new Error(AN_ERROR_OCCURRED)]]);
    expect(mockDisplayError).toHaveBeenCalledTimes(1);

    expect(toJson(wrapper.find('div'))).toMatchSnapshot('jurisdiction error');
  });

  it('renders supersetService error correctly', async () => {
    const mockDisplayError: any = jest.fn();
    (errorHelpers as any).displayError = mockDisplayError;

    const error = new Error('Catastrophy!!');

    const supersetServiceMock: any = jest.fn();
    supersetServiceMock.mockRejectedValue(error);

    const props: JurisdictionMapProps = {
      ...defaultProps,
      jurisdictionId: jurisdiction1.jurisdiction_id,
      supersetService: supersetServiceMock,
    };
    const wrapper = mount(
      <MemoryRouter>
        <JurisdictionMap {...props} />
      </MemoryRouter>
    );

    await flushPromises();
    wrapper.update();

    expect(mockDisplayError.mock.calls).toEqual([[error]]);
    expect(mockDisplayError).toHaveBeenCalledTimes(1);

    expect(toJson(wrapper.find('div'))).toMatchSnapshot('supersetService error');
  });

  it('works when connected to the store', async () => {
    store.dispatch(fetchJurisdictions([jurisdiction1]));

    const supersetServiceMock: any = jest.fn();
    supersetServiceMock.mockImplementation(async () => [jurisdiction1]);

    const props: JurisdictionMapProps = {
        ...defaultProps,
        jurisdictionId: jurisdiction1.jurisdiction_id,
        supersetService: supersetServiceMock,
      } /** the jurisdiction is not supplied as it will be fetched from the store */;

    const wrapper = mount(
      <Provider store={store}>
        <MemoryRouter>
          <ConnectedJurisdictionMap {...props} />
        </MemoryRouter>
      </Provider>
    );

    await flushPromises();
    wrapper.update();

    expect(wrapper.find('GisidaWrapper').props()).toEqual({
      currentGoal: null,
      geoData: jurisdiction1,
      goal: null,
      handlers: [],
      minHeight: '200px',
      pointFeatureCollection: null,
      polygonFeatureCollection: null,
      structures: null,
    });
  });
});
