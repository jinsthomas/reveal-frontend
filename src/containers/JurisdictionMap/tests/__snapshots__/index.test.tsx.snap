// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`containers/JurisdictionMap renders fetch error correctly: jurisdiction error 1`] = `
<div>
  An error prevented this jurisdiction from loading.
</div>
`;
